asset_dashboard/static/js/components/maps/AssetsMap.js

```diff
@@ -0,0 +1,177 @@
+import ReactDOM from 'react-dom'
+import React, { useState, useEffect } from 'react'
+import { GeoJSON } from 'react-leaflet'
+import hash from 'object-hash'
+import Cookies from 'js-cookie'
+import BaseMap from './BaseMap'
+import AreaClipper from '../map_utils/AreaClipper'
+import AssetSearchTable from '../tables/AssetSearchTable'
+
+function AssetTypeOptions() {
+  // these options could come from the server but hardcoding for now 
+  const options = [
+    {value: 'buildings', label: 'Buildings'},
+    {value: 'trails', label: 'Trails'}
+  ]
+
+  return options.map(option => {
+    return <option value={option.value} key={option.value}>{option.label}</option>
+  })
+}
+
+function AssetsMap(props) {
+  const [searchedGeoms, setSearchedGeoms] = useState()
+  const [existingGeoms, setExistingGeoms] = useState()
+  const [clippedGeoms, setClippedGeoms] = useState(null)
+  const [searchText, setSearchText] = useState('')
+  const [searchedAssetType, setSearchedAssetType] = useState('buildings')
+
+  useEffect(() => {
+    if (props?.existing_geoms) {
+      setExistingGeoms(JSON.parse(props.existing_geoms))
+    }
+  }, [setSearchedGeoms, setExistingGeoms])
+
+  function onMapCreated(map) {
+    const group = new L.featureGroup()
+
+    map.eachLayer((layer) => {
+      if (layer.feature) {
+        group.addLayer(layer)
+      }
+    })
+
+    if (Object.keys(group._layers).length > 0) {
+      map.fitBounds(group.getBounds())
+    }
+  }
+
+  function onClipped(featureCollection) {
+    setClippedGeoms(featureCollection)
+  }
+
+  function saveGeometries() {
+    fetch(window.location.pathname, {
+        headers: {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json',
+            'X-CSRFTOKEN': Cookies.get('csrftoken')
+        },
+        method: 'POST',
+        mode: 'same-origin',
+        body: JSON.stringify({
+          'geojson': clippedGeoms,
+          'phase': selectedPhase.value
+        })
+    }).then((response) => {
+      if (response.status == 200) {
+        location.reload()
+        // TODO: show success message
+        // https://stackoverflow.com/questions/13256817/django-how-to-show-messages-under-ajax-function
+      } else {
+        // TODO: catch error and show message
+      }
+    })
+  }
+
+  function searchAssets() {
+    const url = `/assets/?` + new URLSearchParams({
+      'q': searchText,
+      'asset_type': searchedAssetType
+    })
+
+    fetch(url, {
```

Comment:
This request is kind of heavy when there are many results – is there an easy way to show a loading indicator or even "Loading..." text so the user has some feedback that their search is processing?

Reply:
I can just do a simple "Loading..." to start with and we can improve it later.

Reply:
Here's a basic implementation with this commit b397a6. It can definitely be improved but this works for now.
![loading](https://user-images.githubusercontent.com/38969506/150398590-3f9538cd-24ce-4725-93d2-3001181d9276.gif)
9

Reply:
Perf, thanks!